Fixes #33034 - add a description to all provisioning templates

ares opened the pull request:
This PR udpates all the templates metadata to provide at least minimal
description that is then stored in the DB and shown in the edit form.


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ares commented on app/views/unattended/provisioning_templates/PXELinux/windows_default_pxelinux.erb:
```diff
@@ -4,8 +4,10 @@ name: Windows default PXELinux
 - Windows
-%>
+description: |
+  TODO
```
@UXabre any chance you could put together a description for this template?

UXabre replied:
> @UXabre any chance you could put together a description for this template?

A template to render a PXELinux configuration for Windows installations.
It injects the actual installation script from Foreman to give a highly customizable installation.
Note that this does rely on HTTP support inside your PXE environment. If this is not supported, please chain this using iPXE instead.
After this has been injected, it starts wimboot which in turn starts the actual WinPE image

lzap replied:
I think we can just drop a generic one, PXELinux to boot Windows.

Maybe adding a note that Windows boot files are not preconfigured by default by our installer.

ares commented on app/views/unattended/provisioning_templates/iPXE/ipxe_default_local_boot.erb:
```diff
@@ -3,6 +3,8 @@ kind: iPXE
 name: iPXE default local boot
 snippet: false
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Just copy PXELinux local boot description and replace with iPXE. The very same purpose.

ares commented on app/views/unattended/provisioning_templates/iPXE/ipxe_global_default.erb:
```diff
@@ -3,6 +3,8 @@ kind: iPXE
 name: iPXE global default
 snippet: false
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Just copy PXELinux global boot description and replace with iPXE. The very same purpose.

ares commented on app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:
```diff
@@ -3,6 +3,8 @@ kind: iPXE
 name: iPXE intermediate script
 snippet: false
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
This template is returned when `?bootstrap=1` URL parameter is provided.
It is used in pure iPXE environment and it's purpose is to perform another iPXE template call,
but with MAC address(es) of the server in order to find appropriate host record in the inventory.
For more info visit:
https://docs.theforeman.org/nightly/Provisioning_Guide/index-foreman-el.html#Configuring_Networking-Configuring_gPXE_to_Reduce_Provisioning_Times

ares commented on app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb:
```diff
@@ -4,14 +4,15 @@ name: Windows default provision
-- systemTimeZone: GMT Standard Time #see https://msdn.microsoft.com/en-us/library/ms912391(v=winembedded.11).aspx
-- wimImageName = Windows 8.1 Pro # Image name seems only necessary if the WIM contains more than one image
+description: |
+    TODO
```
@UXabre could you please provide a suggestion?

UXabre replied:
> @UXabre could you please provide a suggestion?

A templated answers.xml file for windows installations. This provides all answers to the questions an interactive installation would ask.

lzap replied:
I think it is a "recipe" for some kind of installer (official one)? Make sure to drop the link to the discourse thread (above).

ares commented on app/views/unattended/provisioning_templates/script/grubby_default.erb:
```diff
@@ -3,7 +3,9 @@
 model: ProvisioningTemplate
-%>
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Oh yeah, this is quite rare "script" template kind which is supposed to be just a template you can curl/wget to your server manually (or over ssh). https://projects.theforeman.org/issues/468

Surpringlisly, Grubby is still in EL8 for compatibility reasons: https://access.redhat.com/downloads/content/grubby/8.40-41.el8/x86_64/fd431d51/package

I think it makes sense to move this into ReX template, hey @jlsherrill interested in such patch? Do you still use this? :-)

lzap replied:
Description would be:

Script that can be manually downloaded to a server to reprovision it without booting from network.
Grubby compatibility package must be installed. WARNING: It will immediately reconfigure
bootloader and reprovision the node.
Usage:
  curl https://foreman/unattended/script -o reprovision.sh
  bash reprovision.sh
  reboot

ares commented on app/views/unattended/provisioning_templates/snippet/efibootmgr_netboot.erb:
```diff
@@ -4,7 +4,9 @@ name: efibootmgr_netboot
 snippet: true
-%>
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Most EFI system firmware implementations put new items to the top of the boot priority list. Foreman workflow assumes that managed nodes always boot from network. This snippet can be included from provisioning templates and when "efi_bootentry" host param is set to "previous" it will attempt to find previous boot entry (network) and put it back into first position.

ares commented on app/views/unattended/provisioning_templates/snippet/pxegrub2_chainload.erb:
```diff
@@ -3,7 +3,9 @@ kind: snippet
 snippet: true
-%>
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
In Foreman's typical PXE workflow, managed hosts are configured to always boot from network and inventory build flag dictates if they should boot into installer (build is on) or boot from local drive (build is off). This template is used to chainload from EFI ESP for systems which booted from network. It is not as straightforward as in BIOS and EFI boot file must be found on an ESP partition.

This will only be needed when provisioned hosts are set to boot from network, typically EFI firmware implementations overrides boot order after new OS installation. This behavior can be set in EFI, or "efi_bootentry" host parameter can be set to "previous" to override boot order back to previous (network) setting. See efibootmgr_netboot snippet for more info.

ares commented on app/views/unattended/provisioning_templates/snippet/pxegrub2_discovery.erb:
```diff
@@ -3,13 +3,13 @@ kind: snippet
-below. This will not be fixed until RHEL8 due to prevent breaking changes for
-existing systems. See RHBZ#1259015 for more details.
+description: |
+  TODO
```
@lzap could you please provide a suggestion? The text below does not sound like a full description.

lzap replied:
Snippet with grub2 menu items for discovery plugin. It is included in PXEGrub2 Default templates to boot unknown hosts into discovery image.

ares commented on app/views/unattended/provisioning_templates/snippet/pxegrub2_mac.erb:
```diff
@@ -3,6 +3,8 @@ kind: snippet
 model: ProvisioningTemplate
 snippet: true
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Snippet that is included in all PXEGrub2 templates in order to find MAC-based configuration on TFTP/HTTPBoot. Grub2 in Red Hat is patched with this feature, however, Debian/Ubuntu or other distributions do not have this patch.

ares commented on app/views/unattended/provisioning_templates/snippet/pxegrub_chainload.erb:
```diff
@@ -3,7 +3,9 @@ kind: snippet
 snippet: true
-%>
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Snippet used to chainload BIOS HDD when booted from network. Used in PXEGrub global and local templates.

ares commented on app/views/unattended/provisioning_templates/snippet/pxegrub_discovery.erb:
```diff
@@ -3,13 +3,13 @@ kind: snippet
-below. This will not be fixed until RHEL8 due to prevent breaking changes for
-existing systems. See RHBZ#1259015 for more details.
+description: |
+  TODO
```
@lzap could you please provide a suggestion? The text below does not sound like a full description.

lzap replied:
Snippet with grub menu items for discovery plugin. It is included in PXEGrub Default templates to boot unknown hosts into discovery image.

ares commented on app/views/unattended/provisioning_templates/snippet/pxelinux_chainload.erb:
```diff
@@ -3,7 +3,9 @@ kind: snippet
 snippet: true
-%>
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Snippet used to chainload BIOS HDD when booted from network. Used in PXELinux global and local templates.

ares commented on app/views/unattended/provisioning_templates/snippet/pxelinux_discovery.erb:
```diff
@@ -3,11 +3,11 @@ kind: snippet
-In short, before RHEL 7.2 virtio driver didn't have the new persistent naming scheme, causing interfaces to be named eth0, eth1, etc..
-If you want to enable the new persistent naming scheme and get inteface names like ens3, add net.ifnames=1 to the APPEND line below.
+description: |
+  TODO
```
@lzap could you please provide a suggestion?

lzap replied:
Snippet with PXELinux menu items for discovery plugin. It is included in PXELinux Default templates to boot unknown hosts into discovery image.

ares commented on app/views/unattended/provisioning_templates/snippet/saltstack_minion.erb:
```diff
@@ -3,7 +3,9 @@ kind: snippet
 snippet: true
-%>
+description: |
+  TODO
```
@sbernhard do you know who could provide a suggestion what to put in here?

ares commented on app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb:
```diff
@@ -9,8 +9,14 @@ oses:
 - Rocky
 - Debian
 - Ubuntu
+description: |
```
@timogoebel could you please take a look and see if I covered that correctly? You also mention a finish template or cloud-init could be used after this, however is that possible today with Foreman? I think we can't chain multiple user-data templates for single OS nor we can combine finish and user-data. Also if you have any other comments for templates you know well, please add that. Thanks!

timogoebel replied:
Yes, we can use this mechanism and cloud-init combined. See https://github.com/theforeman/foreman_userdata (which is now part of core).
Note, that is is required as vsphere does not support running a finish script as part of the customization spec.

lzap replied:
See https://docs.theforeman.org/nightly/Provisioning_Guide/index-foreman-el.html#Provisioning_Virtual_Machines_in_VMware_vSphere-Provisioning_with_cloudinit_and_userdata_templates

ares committed:
Fixes #33034 - add a description to all provisioning templates

This PR udpates all the templates metadata to provide at least minimal
description that is then stored in the DB and shown in the edit form.

Co-authored-by: Timo Goebel <mail@timogoebel.name>
Co-authored-by: Lukas Zapletal <lzap@redhat.com>

timogoebel commented on app/views/unattended/provisioning_templates/user_data/userdata_open_vm_tools.erb:
```diff
@@ -9,8 +9,14 @@ oses:
+description: |
+  Template for VMWare customization via open-vm-tools during user-data image based provisioning.
+
+  Comparing to other user-data templates e.g. those based on the cloud-init, the open-vm-tools
+  allows to configure networking of the VM prior boot, so no DHCP configuration is required.
+  Cloud-init is typically used for a stage 2 configuration. In order for this to work, the VMware
+  template must have open-vm-tools installed.
```
```suggestion
  Comparing to other user-data templates, e.g. those based on cloud-init, this template defines a Customization Specifications that is evaluated by vSphere while cloning a VM from a Template. When the Template that should be cloned has open-vm-tools installed, this allows to configure certain aspects, e.g. networking of the VM, prior boot, so no DHCP configuration is required.
  Cloud-init is typically used for a stage 2 configuration. In order for this to work, the VMware
  template must have open-vm-tools installed.
  The example below is for Linux only, Windows is supported by this mechanism via Sysprep as well, though.
  
  This needs to he a hash in YAML format with these keys:
  - encryptionKey (array)
  - globalIPSettings (hash), REQUIRED
  - identity (hash), REQUIRED
  - nicSettingMap (array)
  - options (hash)
  - extraConfig (hash)
  
  See https://vdc-repo.vmware.com/vmwb-repository/dcr-public/6b586ed2-655c-49d9-9029-bc416323cb22/fa0b429a-a695-4c11-b7d2-2cbc284049dc/doc/vim.vm.customization.Specification.html
```

lzap commented on app/views/unattended/provisioning_templates/PXEGrub/jumpstart_default_pxegrub.erb:
```diff
@@ -4,7 +4,10 @@ name: Jumpstart default PXEGrub
+description: |
+  The template to render PXEGrub bootloader configuration for Solaris.
+  The output is deployed on the host's subnet TFTP proxy.
+-%>
```
Maybe adding something like this to all PXEGrub templates:

This template is for legacy Grub 1.x, you are probably looking for Grub2.

lzap commented on app/views/unattended/provisioning_templates/PXEGrub/jumpstart_default_pxegrub.erb:
```diff
@@ -4,7 +4,10 @@ name: Jumpstart default PXEGrub
 - Solaris
-%>
+description: |
+  The template to render PXEGrub bootloader configuration for Solaris.
```
Nitpick: The software is named Grub (version 1 or 2). Historically, Foreman had template kind named "PXELinux" so we continued with this with "PXEGrub" and "PXEGrub2". But technically it is incorrect, nothing is named like this. We could use the correct term in all descriptions, bonus: it is shorter.

lzap commented on app/views/unattended/provisioning_templates/PXELinux/coreos_pxelinux.erb:
```diff
@@ -4,7 +4,11 @@ name: CoreOS PXELinux
+  The template to render PXELinux bootloader configuration for CoreOS Container Linux distribution.
+  This does not work for Fedora CoreOS or Flatcar Container Linux, which use Ignition instead
+  The output is deployed on the host's subnet TFTP proxy.
+-%>
```
Since CoreOS Container Linux is EOL shall we drop a note here? Just thinking, perhaps not. Maybe to remove it later (Deprecation).

ares replied:
I keep the list I want to propose for dropping later after this is in, so far:

* coreos (replaced by fedora/flatcar),
* poap,
* RancherOS PXELinux (replaced by k3os),
* WAIK (likely replaced by other windows templates), 
* Atomic Kickstart default (is there still support for RH Atomic?), 
* CoreOS provision, 
* rancher os and coreos snippets (content of their provisioning templates), 
* http_proxy - unused snippet

but I'll open proper discourse thread for that

lzap commented on app/views/unattended/provisioning_templates/PXELinux/kickstart_ovirt_pxelinux.erb:
```diff
@@ -5,7 +5,10 @@ model: ProvisioningTemplate
 - RHVH
-%>
+description: |
+  The template to render PXELinux bootloader configuration for oVirt and Red Hat Virtualization Host distributions.
```
```suggestion
  The template to render PXELinux bootloader configuration for oVirt Node or Red Hat Virtualization Host.
```

lzap commented on app/views/unattended/provisioning_templates/PXELinux/pxelinux_chain_ipxe.erb:
```diff
@@ -2,15 +2,14 @@
-http://projects.theforeman.org/projects/foreman/wiki/Fetch_boot_files_via_http_instead_of_TFTP
-%>
+  For more info visit:
+  http://projects.theforeman.org/projects/foreman/wiki/Fetch_boot_files_via_http_instead_of_TFTP
```
```suggestion
  https://docs.theforeman.org/nightly/Provisioning_Guide/index-foreman-el.html#Configuring_Networking-Configuring_gPXE_to_Reduce_Provisioning_Times
```

lzap commented on app/views/unattended/provisioning_templates/PXELinux/pxelinux_chain_ipxe_undi.erb:
```diff
@@ -2,17 +2,16 @@
-http://projects.theforeman.org/projects/foreman/wiki/Fetch_boot_files_via_http_instead_of_TFTP
-%>
+  For more info visit:
+  http://projects.theforeman.org/projects/foreman/wiki/Fetch_boot_files_via_http_instead_of_TFTP
```
```suggestion
  https://docs.theforeman.org/nightly/Provisioning_Guide/index-foreman-el.html#Configuring_Networking-Configuring_gPXE_to_Reduce_Provisioning_Times
```

lzap commented on app/views/unattended/provisioning_templates/finish/windows_default_finish.erb:
```diff
@@ -8,23 +8,25 @@ oses:
-%>
+description: |
+  A finish template executed at the end of Windows provisioning. For more information, please
+  see https://community.theforeman.org/t/windows-provisioning-made-easy/16756/87
```
```suggestion
  see https://community.theforeman.org/t/windows-provisioning-made-easy/16756
```

We can probably drop the same link for all "Windows Default *" templates.

lzap commented on app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -9,6 +9,10 @@ oses:
 - Rocky
+description: |
+  The template to render iPXE installation script for kickstart based distributions
+  The output is deployed on the host's subnet TFTP proxy.
```
I just realized that this should be

```suggestion
  The output is deployed on the host's subnet TFTP or HTTPBoot proxy.
```

pretty much everywhere.

lzap commented on app/views/unattended/provisioning_templates/provision/kickstart_ovirt.erb:
```diff
@@ -5,37 +5,38 @@ model: ProvisioningTemplate
-- ntp-server: string (default="0.fedora.pool.ntp.org")
-- disable-firewall: boolean (default=false)
+description: |
+  Provisioning template for oVirt and Red Hat Virtualization Host distributions. It renders the kickstart
```
```suggestion
  Provisioning template for oVirt Node and Red Hat Virtualization Host. It renders the kickstart
```

lzap commented on app/views/unattended/provisioning_templates/script/windows_default_script.erb:
```diff
@@ -4,13 +4,12 @@ name: Windows peSetup.cmd
+  You can create and assign a "user_data" ProvisioningTemplate as powershell to execute some custom code
+  params:
+  - wimImageName: Windows 8.1 Pro # name of wim image to apply
+-%>
```
Add a link to the discourse thread.

lzap commented on app/views/unattended/provisioning_templates/snippet/bmc_nic_setup.erb:
```diff
@@ -3,42 +3,40 @@ kind: snippet
+  How to test configuration result:
+  * From a remote host that can reach the current IPMI interface, install ipmitool utilities and type the following command:
+    ipmitool -I <interface> -L <privilege_level> -U <user> -H <host> chassis status
+-%>
```
If this is not in our docs, this could be a good contribution.

ares replied:
where do I open tickets for that? :-) I really can't document this myself

lzap commented on app/views/unattended/provisioning_templates/snippet/create_users.erb:
```diff
@@ -3,6 +3,8 @@ name: create_users
 kind: snippet
 snippet: true
+description: |
+  This snippet can be used  
```
Absolutely :-D

However, it looks like it is not used anywhere in templates.

lzap replied:
Maybe this is used by ReX @adamruzicka ? Can you provide description in that case?

ares replied:
I've finished the description, it's quite clear :-)

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -3,6 +3,10 @@ kind: snippet
 model: ProvisioningTemplate
 snippet: true
+description: |
+  The list of kernel options for initrd appended to the bootloader kernel line.
```
```suggestion
  The list of kernel options for initrd appended to the bootloader kernel line on Red Hat compatible distributions.
```

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_networking_setup.erb:
```diff
@@ -3,6 +3,9 @@ name: kickstart_networking_setup
 kind: snippet
+description: |
+  Generates a complete shell script that creates ifcfg configuration files for all
+  host's network interfaces. Typically used on RPM based distributions.
```
```suggestion
  host's network interfaces. Typically used on Red Hat based distributions.
```

SLES uses RPMs but this wont't probably work I think.

lzap commented:
I am done, nice, well done, have you considered career in our provisioning team? :-)

Please don't make me to read this again, just ping me if there is any additional question and let's gather feedback from others and merge it. Nice contribution.

ares commented:
Squashed all suggestions, resolved comments which I didn't explicitly replied to and rebased. There's still one for salt minion I need a description for, ping @sbernhard 

I'd like to get this in 3.1. Next steps (as a reminder)
* drop unused templates
* parse the metadata out of the template so it does not take the space in the editor field
* continuously improve :-)

ares commented:
Ok, I figured out the last one based on the salt documentation. @lzap I think this is now ready for merge, snapshots should be updated.

lzap reviewed (approved):
Thanks merging, that was a ton of work!
